Archive dev3 - Adding request handlers for data archival process.

gavaletz opened the pull request:
Here are the handlers to make use of the utility functions that were added late last week.  There are no unit tests on these handlers as it would only be testing things like the datastore and APIs that should be tested internally.  To try this out please use the instance at http://20120429-172305-gavaletz-94a9133.openmobiledata.appspot.com/.

I am waiting on an update from Dave on a group bucket to use.  Once that is sorted out there will be some decisions to be made on ACLs.  Since the amount of data in the datastore is pretty limited at this point, these requests are returning fine from our frontend instances, but we will want to configure these requests as backend jobs in the near future.  Another thing that we can do to reduce the load (noted as a TODO in the code) is to create an entity with the archive parameters so that when similar requests are made we can simply return the link to the pre-made file in Google Storage and not repeat our work.  Once we start keeping track of these it will be important to have the hashes for these files as well.

gavaletz committed:
Adding scripts for getting data into dev_appserver.

These changes do not change the way the application runs on GAE but are
necessary for locally testing operations that operate on data.

server/download_dev_data.sh:
There is a documented procedure to download data from the datastore
hosted on GAE to a local SQLite3 file using the appcfg.py utility, but
it is a little cumbersome so why not sutomate it.  This does not pruduce
data that the dev_appserver can use so we need to "upload" the data we
just downloaded into the dev_appserver.

server/upload_local_data.sh:
This basically puts the data into the datastore that is referenced by
the dev_appserver.  To make use of this you will need to run the local
instance and make sure that you are using the lines for large data
operations.  This will ensure that you don't have to do this twice.

server/script_config.sh:
When writing the scripts for downloading and uploading data there were a
lot of shared values and configuring things was becoming confusing.
Factored out the variables into one file that can be sourced into the
other scripts.

server/update_dev_server.sh:
Changed due to refactoring above.

server/run_server_locally.sh:
Changed due to refactoring above, and added some configuration to limit
the amount of debuging messages displayed and sourcing the data
downloaded from GAE.

gavaletz committed:
Responding to feedback for commit 365a6bbb5bf8.

Fixed typo in upload_local_data.sh, removed line from
download_dev_data.sh that made a backup copy of the downloaded data, and
gave the user warning that we would be writing to a few files.  This is
also an attempt to try rebase instead of pull from master.  The former
seems to be a bit noisy.

gavaletz committed:
Small bugfix for commit 365a6bbb5bf8.

server/upload_local_data.sh:
We should be reading from the same file that the
server/download_dev_data.sh script saved the data in.  Changed
$DATA_PATH to $DOWNLOAD_DATA_PATH.

gavaletz committed:
Adding utility functions for data archival process.

These files will be used to serialize and compress entities from the
datastore on GAE into a zip file.  Handlers for triggering these
functions will be in the next CL.

server/gspeedometer/helpers/archive_util.py - see file commentary for
details on what functions have been added.

server/gspeedometer/helpers/util.py - added a couple comments about
issues that need to be addressed.  Issues will be added in the github
issue tracker.  One problem is that it is possible to have dangling
references when some data is removed and ConvertToDict will throw a
ReferencePropertyResolveError.  The other problem is that
ConvertFromDict does not seem to be able to deal with unconverting
output from ConvertToDict in some cases.

server/gspeedometer/model.py - added a function that will take a string
and return the class that is represented by that string.  This avoids
evaluation of the string and the possible security issues that brings.

gavaletz committed:
Responding to feedback for commit e27c1fa50342.

The big change here is that for now we are going to support only the
export of measurement types.  This reduces the code base considerably and
takes advantage of more existing code.

server/gspeedometer/controllers/measurement.py - code for taking a list
of measurement entities and changing them into a list of dictionaries
has been moved to server/gspeedometer/helpers/util.py.  This is needed
since we would like to support more variety in how we obtain a list of
measurements.

server/gspeedometer/helpers/archive_util.py - removed methods needed for
importing data and the ArchivePack method.  The MeasuermentListToDictList
(taken from server/gspeedometer/controllers/measurement.py) will be used
over the ArchivePack method.

server/gspeedometer/helpers/archive_util_test.py - unittests for the
methods added with server/gspeedometer/helpers/archive_util.py.

server/gspeedometer/helpers/util.py - code added from
server/gspeedometer/controllers/measurement.py.

server/gspeedometer/model.py - removed unneeded method for looking up a
class by string.

gavaletz committed:
Adding request handlers for data archival process.

These changes will allow measurement data from the datastore to be
downloaded as a zip file, or posted to Google Storage for Developers as
a zip file.  The data can be grouped my device, date or both.  Having
the ability to archive by device allows for any user to see his data and
would allow something akin to the "takeout" service that Google
provides.  The drawback is that this requires another index to be
maintained and increases the overall costs for writes to the datastore.

server/app.yaml.tmpl - adding the requirement that requests made to any
/admin/* URL can only be made by those with admin access to the GAE
instance.  This is sufficient since anyone with GAE admin access would
have access to the whole datastore anyways.

server/gspeedometer/config.py - adding configuration points for Google
Storage for Developers.

server/gspeedometer/controllers/archive.py - adding request handlers for
downloading data and sending it to Google Storage for Developers.  These
requests would normally be handled by backends and would be used with a
chron job.

server/gspeedometer/helpers/util.py - fixed a copy/paste refactoring
problem from commit c40a958e5cea.

server/gspeedometer/main.py - adding our new handlers to the django
mapping.  See the changes to server/app.yaml.tmpl as well for security
implications.

server/index.yaml - new index added to allow for filtering measurements
by device_id.

gavaletz commented:
I hope that the diff on this will update when the CL from archive_dev2 gets merged.  It is really not that big, but just builds on top of the changes in archive_dev2.

mdwelsh commented on server/gspeedometer/config.py:
```diff
@@ -61,3 +61,9 @@
 ACTIVE_DAYS = 5
+
+#Archive Settings
+CONTENT_TYPE = 'application/zip'
```
Maybe ARCHIVE_CONTENT_TYPE just to be unambiguous.

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/config.py:
```diff
@@ -61,3 +61,9 @@
+
+#Archive Settings
+CONTENT_TYPE = 'application/zip'
+CONTENT_DISPOSITION_BASE = 'attachment; filename="%s.zip"'
```
Add ARCHIVE_ to each of these.

gavaletz replied:
Just checking these off as I go.  Done

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+  if end:
+    measurement_q.filter('timestamp <', end)
+  measurement_q.order('timestamp')
+  measurement_list = measurement_q.fetch(config.QUERY_FETCH_LIMIT)
```
Hmm, I wonder if you don't just want to return the query and let the caller fetch the results and turn them into a list? Again I'm really really worried about the memory consumption of loading a bunch of measurements into memory all at once. Also, a caller of this method would have no way of knowing that only QUERY_FETCH_LIMIT results were returned when there may have been others - I'd prefer the caller control how fetching is done.

gavaletz replied:
I agree, but this will bring other changes.  Marked as a todo.

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+    model_list = GetMeasurementDictList(device_id, start, end)
+
+    # Serialize the data
+    data = {'Measurement': json.dumps(model_list)}
```
Here's a case where you could iterate over results from the query returned by GetMeasurementDictList (renamed to GetMeasurementDictQuery) and serialize them one by one. We would eventually change this code to construct the archive zip files on the fly rather than loading everything into memory and writing it out, but at least with that small change that will be easy to do later.

gavaletz replied:
This should happen with a change to MeasurementListToDictList since we are using that for part of the serialization.  That was the code that came from controllers/measurement.py (you mentioned wanting to improve this).  I think that if we can change helpers/util.py:MeasurementListToDictList to use a query and iterate over it as you suggested, then the controllers/archive.py:GetMeasurementDictList could take a more proactive approach to dealing with large query results.

Again marking with a TODO for now.

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+    # Serialize the data
+    data = {'Measurement': json.dumps(model_list)}
+
+    # Generate directory/file name based on parameters
```
Would be nice to factor this out into its own method allowing you to write a unit test for it.

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+  that can be returned to the user or stored in BigStore.
+  """
+
+  def __Archive(self, **unused_args):
```
_Archive rather than __Archive.

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+    try:
+      archive_dir, archive_data = self.__Archive()
+      self.response.headers['Content-Type'] = config.CONTENT_TYPE
+      cd = config.CONTENT_DISPOSITION_BASE % archive_dir  # Stupid line length.
```
You can use ( ... ) with a single item (not a tuple) to deal with the line length issue. Also, please do not editorialize in comments.

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+      logging.exception(e)
+      self.response.clear()
+      self.response.set_status(500)
+      self.response.out.write('Try doing this on a backend instance...')
```
More useful to include the full error text and a more descriptive error message to help with debugging.

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+    # Consider saving the file to GS too so that it can be returned from there
+    # if requested again in the future.
+
+  def ArchiveToGs(self, **unused_args):
```
Change to ArchiveToGoogleStorage -- "Gs" is not self-explanatory.

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+      logging.exception(e)
+      self.response.clear()
+      self.response.set_status(500)
+      self.response.out.write('Try doing this on a backend instance...')
```
As above, please use more descriptive error messages.

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+
+      # Finalize (a special close) the file.
+      files.finalize(gs_archive)
+      self.response.out.write('OK -- %s' % gs_archive_name)
```
To be consistent with the rest of this app, please use a JSON-encoded response, e.g., { status: "OK", archive_name: "gs_archive_name" }

gavaletz replied:
Done, also converted the error responses to a JSON-encoded response.

mdwelsh commented on server/gspeedometer/helpers/archive_util.py:
```diff
@@ -0,0 +1,95 @@
+    archive_file = zipfile.ZipFile(file=archive_stream,
+      compression=zipfile.ZIP_DEFLATED, mode='w')
+  except RuntimeError, e:
+    logging.exception('likely missing the zlib module: %s', e)
```
Why would this happen? (Would this happen on GAE?)

mdwelsh replied:
Also, don't need %s in the string - passing the exception as the second parameter to logging.exception does the right thing.

gavaletz replied:
%s removed.

Re the exception -- It is unlikely to happen.  It's the exceptional case :-)

It is easy to catch this case so why not?  I came across documentation stating why a generic runtime exception would be thrown in each of these cases, and if the problem ever comes up. the person looking back to the code will know.  Otherwise it isn't clear what caused the generic RuntimeError.  Simple enough to remove this, but insignificant in cost so my vote is to keep it, but you can always veto.

mdwelsh commented on server/gspeedometer/helpers/archive_util.py:
```diff
@@ -0,0 +1,95 @@
+    try:
+      archive_file.writestr(info, file_dict[member_name])
+    except RuntimeError, e:
+      logging.exception('likely FakeFile is closed or \'r\': %s', e)
```
You don't need the %s in the string - passing the exception as the second parameter to logging.exception does the right thing.

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/helpers/util.py:
```diff
@@ -133,3 +133,59 @@ def ConvertFromDict(model, input_dict, include_fields=None,
     else:
       setattr(model, k, v)
+
+def MeasuermentListToDictList(measurement_list, include_fields=None,
```
Typo: "Measurement". You have this typo in several places, including the comments, please scour the source for this.

mdwelsh replied:
Should this method belong in model.py?

mdwelsh replied:
This method needs a unit test.

gavaletz replied:
Typo fixed and added a todo(mdw) for the unit test (you can reassign this later).

I thing the this method is a closer fit to the other conversion methods in helpers/util.py as things are right now.  The only things in model.py are the actual model classes.  I think that if the number of conversion methods grows, then it might make sense to have a model_util.py.

mdwelsh commented on server/gspeedometer/helpers/util.py:
```diff
@@ -133,3 +133,59 @@ def ConvertFromDict(model, input_dict, include_fields=None,
+
+  Given a list of measuerment model objects from the datastore, this method
+  will convert that list into a list of python dictionaries that can then
+  be serailized.
```
Typo: serialized

You might want to turn on spellcheck in your editor :-)

gavaletz replied:
Done

mdwelsh commented on server/gspeedometer/controllers/archive.py:
```diff
@@ -0,0 +1,230 @@
+# Copyright 2012 Google Inc.
```
All of these methods need unit tests. They are not hard to write, so I am not sure why you think we should not be doing them - in general I would like to see unit tests for all new code.

gavaletz replied:
Added a todo(mdw) for the unit test (either I will get to some of these next week, or you can reassign this later).

gavaletz committed:
Responding to feedback on commit f6d36a9e12bb.

/server/gspeedometer/config.py - made archive variables more obvious.

/server/gspeedometer/controllers/archive.py - re-factored the method to
create a file name based on parameters.  Fixed some typos, and renaming
of variables.

/server/gspeedometer/controllers/archive_test.py - unit test for the
method to create a file name based on parameters.

/server/gspeedometer/controllers/measurement.py - fixed typo and added
some TODOs.

/server/gspeedometer/helpers/archive_util.py - fixed some syntax issues
and renaming of variables.

/server/gspeedometer/helpers/util.py - fixed typos.

/server/gspeedometer/main.py - fixed renaming of variables.
